src/spi.rs

```diff
@@ -1641,68 +1654,160 @@ fn spi_create_transaction(
     #[cfg(not(esp_idf_version = "4.3"))]
     if _keep_cs_active {
         transaction.flags |= SPI_TRANS_CS_KEEP_ACTIVE
     }
+}
 
+fn spi_transmit(
+    handle: spi_device_handle_t,
+    transactions: impl Iterator<Item = spi_transaction_t>,
+    polling: bool,
+    queue_size: usize,
+) -> Result<(), EspError> {
     if polling {
-        esp!(unsafe { spi_device_polling_transmit(handle, transaction as *mut _) })
+        for mut transaction in transactions {
+            esp!(unsafe { spi_device_polling_transmit(handle, &mut transaction as *mut _) })?;
+        }
     } else {
-        esp!(unsafe { spi_device_transmit(handle, transaction as *mut _) })
+        let mut transaction_queue = [spi_transaction_t::default(); MAX_QUEUED_TRANSACTIONS];
+        let mut queue_iter = transaction_queue.iter_mut().take(queue_size);
+
+        for transaction in transactions {
+            let slot = queue_iter.next();
+            let trans = if let Some(slot) = slot {
+                slot
+            } else {
+                // If the queue is full, we wait for the first transaction in the queue and use it
+                // for the next one.
+                let mut ret_trans: *mut spi_transaction_t = ptr::null_mut();
+                esp!(unsafe {
+                    spi_device_get_trans_result(handle, &mut ret_trans as *mut _, delay::BLOCK)
+                })?;
+                unsafe { &mut *ret_trans }
+            };
+
+            // Write transaction to stable memory location
+            *trans = transaction;
+            esp!(unsafe { spi_device_queue_trans(handle, trans as *mut _, delay::BLOCK) })?;
+        }
+
+        let queued_transactions = min(MAX_QUEUED_TRANSACTIONS, queue_size) - queue_iter.len();
+        for _ in 0..queued_transactions {
+            let mut ret_trans: *mut spi_transaction_t = ptr::null_mut();
+            esp!(unsafe {
+                spi_device_get_trans_result(handle, &mut ret_trans as *mut _, delay::BLOCK)
+            })?;
+        }
     }
+    Ok(())
 }
 
 async fn spi_transmit_async(
     handle: spi_device_handle_t,
-    transaction: &mut spi_transaction_t,
-    _keep_cs_active: bool,
+    transactions: impl Iterator<Item = spi_transaction_t>,
+    queue_size: usize,
 ) -> Result<(), EspError> {
-    // This unfortunately means that this implementation is incorrect for esp-idf < 4.4.
-    // The CS pin should be kept active through transactions.
-    #[cfg(not(esp_idf_version = "4.3"))]
-    if _keep_cs_active {
-        transaction.flags |= SPI_TRANS_CS_KEEP_ACTIVE
-    }
-
-    let notification = Notification::new();
-
-    transaction.user = ptr::addr_of!(notification) as *mut _;
-
-    match esp!(unsafe { spi_device_queue_trans(handle, transaction as *mut _, delay::NON_BLOCK) }) {
-        Err(e) if e.code() == ESP_ERR_TIMEOUT => unreachable!(),
-        other => other,
-    }?;
+    let notifications: [Notification; MAX_QUEUED_TRANSACTIONS] = [
```

Comment:
Nit: you can use [this trick](https://github.com/esp-rs/esp-idf-hal/blob/master/src/gpio.rs#L1654) to avoid enumerating all the 10 notifications explicitly.

Reply:
Interesting, I'll do it.